src/operator/deinput.rs

```diff
@@ -0,0 +1,703 @@
+
+    /// Reserve space for at least `reservation` more updates in the
+    /// internal input buffer.
+    ///
+    /// Reservations are not required but can be used when the number
+    /// of inputs is known ahead of time to reduce reallocations.
+    fn reserve(&mut self, reservation: usize);
+
+    /// Push all buffered updates to the underlying input stream handle.
+    ///
+    /// Flushed updates will be pushed to the stream during the next call
+    /// to [`DBSPHandle::step`](`crate::DBSPHandle::step`).  `flush` can
+    /// be called multiple times between two subsequent `step`s.  Every
+    /// `flush` call adds new updates to the previously flushed updates.
+    ///
+    /// Updates queued after the last `flush` remain buffered in the handle
+    /// until the next `flush` or `clear_buffer` call or until the handle
+    /// is destroyed.
+    fn flush(&mut self);
+
+    /// Clear all buffered updates.
+    ///
+    /// Clears updates pushed to the handle after the last `flush`.
+    /// Flushed updates remain queued at the underlying input handle.
+    // TODO: add another method to invoke `CollectionHandle::clear_input`?
+    fn clear_buffer(&mut self);
+
+    /// Create a new handle connected to the same input stream.
+    ///
+    /// The new handle will use its own input buffer, but shares the
+    /// underlying input stream handle with the original handle.
+    fn fork(&self) -> Box<dyn DeCollectionHandle>;
+}
+
+/// An input handle that wraps a [`CollectionHandle<V, R>`](`CollectionHandle`)
+/// returned by [`Circuit::add_input_zset`](`crate::Circuit::add_input_zset`).
+///
+/// The [`insert`](`Self::insert`) method of this handle deserializes value
+/// `v` type `V` and buffers a `(v, +1)` update for the underlying
+/// `CollectionHandle`.
+///
+/// The [`delete`](`Self::delete`) method of this handle deserializes value
+/// `v` type `V` and buffers a `(v, -1)` update for the underlying
+/// `CollectionHandle`.
+pub struct DeZSetHandle<K, R> {
+    updates: Vec<(K, R)>,
+    handle: CollectionHandle<K, R>,
+}
+
+impl<K, R> DeZSetHandle<K, R> {
+    pub fn new(handle: CollectionHandle<K, R>) -> Self {
+        Self {
+            updates: vec![],
+            handle,
+        }
+    }
+
+    fn clear(&mut self) {
+        if self.updates.capacity() > MAX_REUSABLE_CAPACITY {
+            self.updates = vec![];
+        }
+    }
+}
+
+impl<K, R> DeCollectionHandle for DeZSetHandle<K, R>
+where
+    K: DBData + for<'de> Deserialize<'de>,
+    R: DBWeight + ZRingValue,
+{
+    fn insert(&mut self, deserializer: &mut dyn ErasedDeserializer) -> Result<(), EError> {
+        let key = deserialize::<K>(deserializer)?;
+
+        self.updates.push((key, R::one()));
+        Ok(())
+    }
+
+    fn delete(&mut self, deserializer: &mut dyn ErasedDeserializer) -> Result<(), EError> {
+        let key = deserialize::<K>(deserializer)?;
+
+        self.updates.push((key, R::one().neg()));
```

Comment:
can you just do this to a collection? I understand for a zset

Reply:
This is a Z-set handle, so this is well-defined.